src/Passes/Source/ModuleLoader/ModuleLoader.hpp

```diff
@@ -0,0 +1,132 @@
+        using FunctionAnalysisManager = llvm::FunctionAnalysisManager;
+
+        explicit SingleModuleTransformation(
+            OptimizationLevel const& optimisation_level = OptimizationLevel::O0,
+            bool                     debug              = false)
+          : loop_analysis_manager_{debug}
+          , function_analysis_manager_{debug}
+          , gscc_analysis_manager_{debug}
+          , module_analysis_manager_{debug}
+          , optimisation_level_{optimisation_level}
+          , debug_{debug}
+        {
+
+            pass_builder_.registerModuleAnalyses(module_analysis_manager_);
+            pass_builder_.registerCGSCCAnalyses(gscc_analysis_manager_);
+            pass_builder_.registerFunctionAnalyses(function_analysis_manager_);
+            pass_builder_.registerLoopAnalyses(loop_analysis_manager_);
+
+            pass_builder_.crossRegisterProxies(
+                loop_analysis_manager_, function_analysis_manager_, gscc_analysis_manager_, module_analysis_manager_);
+
+            module_pass_manager_.addPass(RemoveDisallowedAttribubtesPass());
+        }
+
+        bool apply(llvm::Module* module)
+        {
+            module_pass_manager_.run(*module, module_analysis_manager_);
+
+            if (verifyModule(*module, &llvm::errs()))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        bool isDebugMode() const
+        {
+            return debug_;
+        }
+
+      private:
+        llvm::PassBuilder             pass_builder_;
+        llvm::LoopAnalysisManager     loop_analysis_manager_;
+        llvm::FunctionAnalysisManager function_analysis_manager_;
+        llvm::CGSCCAnalysisManager    gscc_analysis_manager_;
+        llvm::ModuleAnalysisManager   module_analysis_manager_;
+
+        llvm::ModulePassManager module_pass_manager_{};
+        OptimizationLevel       optimisation_level_{};
+        bool                    debug_{false};
+    };
+
+    class ModuleLoader
+    {
+      public:
+        using Module       = llvm::Module;
+        using Linker       = llvm::Linker;
+        using SMDiagnostic = llvm::SMDiagnostic;
+
+        explicit ModuleLoader(Module* final_module)
+          : final_module_{final_module}
+          , linker_{*final_module}
+
+        {
+        }
+
+        bool addModule(std::unique_ptr<Module>&& module, String const& filename = "unknown")
+        {
+            if (verifyModule(*module, &llvm::errs()))
+            {
+                llvm::errs() << filename << ": "
+                             << "input module is broken!\n";
+                return false;
+            }
+
+            return !linker_.linkInModule(std::move(module), Linker::Flags::None);
+        }
+
+        bool addIrFile(String const& filename)
+        {
+
+            // Loading module
+            SMDiagnostic            err;
+            std::unique_ptr<Module> module = parseIRFile(filename, err, final_module_->getContext());
```

Comment:
I fail to find the definition of `parseIRFile()`.

Reply:
It is an LLVM function: https://llvm.org/doxygen/IRReader_2IRReader_8h.html

Reply:
(Just wanted to make sure that the second sentence is not forgotten)